clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java

```diff
@@ -116,9 +116,9 @@
     public void seekToEnd(Collection<TopicPartition> partitions);
 
     /**
-     * @see KafkaConsumer#position(TopicPartition)
+     * @see KafkaConsumer#position(TopicPartition, long)
      */
-    public long position(TopicPartition partition);
+    public long position(TopicPartition partition, long timeout);
```

Comment:
I think the root cause of the hanging is that, the `list offset` response returns empty indicating the broker did not know about this topic, and the client will hence retry forever. So what I suggest is that just making the metadata refresh once in `position()`. As for the flaky unit tests @baluchicken , we can use waitForConditions in those test rather than block waiting until it succeeds.